### PR TITLE
[UrlConnectionHttpClient] Wrap erroneous NPE from HttpURLConnection as IOException

### PR DESCRIPTION
**Pull request by Bennett-Lynch**
## Description

* When interacting specifically with HttpURLConnection#getResponseCode,
wrap any unexpected NullPointerExceptions as IOException
* Add corresponding unit test

## Motivation and Context

`HttpURLConnection#getInputStream0` has been observed to intermittently
throw `NullPointerException`s for reasons that still require further
investigation, but are assumed to be due to a bug in the JDK.
Propagating such NPEs is confusing for users and are not subject to
being retried on by the default retry policy configuration, so instead
we should bias towards propagating these as `IOException`s.

## Testing

Created new unit test to simulate the observed (but not yet reliably reproduced) behavior of HttpURLConnection throwing NPE.

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)

## Checklist

- [x] I have read the **CONTRIBUTING** document
- [x] Local run of `mvn install` succeeds
- [x] My code follows the code style of this project
- [ ] My change requires a change to the Javadoc documentation
- [ ] I have updated the Javadoc documentation accordingly
- [x] I have read the **README** document
- [x] I have added tests to cover my changes
- [x] All new and existing tests passed
- [x] A short description of the change has been added to the **CHANGELOG**
- [ ] My change is to implement 1.11 parity feature and I have updated [LaunchChangelog](https://github.com/aws/aws-sdk-java-v2/blob/master/docs/LaunchChangelog.md)

## License

- [x] I confirm that this pull request can be released under the Apache 2 license

### CODE REVIEW

File: http-clients/url-connection-client/src/main/java/software/amazon/awssdk/http/urlconnection/UrlConnectionHttpClient.java [hunk 1]
```diff
@@ -224,6 +224,21 @@ public HttpExecuteResponse call() throws IOException {
                                       .build();
         }
 
+        /**
+         * {@link sun.net.www.protocol.http.HttpURLConnection#getInputStream0()} has been observed to intermittently throw
+         * {@link NullPointerException}s for reasons that still require further investigation, but are assumed to be due to a
+         * bug in the JDK. Propagating such NPEs is confusing for users and are not subject to being retried on by the default 
+         * retry policy configuration, so instead we bias towards propagating these as {@link IOException}s.
+         */
+        private static int getResponseCodeSafely(HttpURLConnection connection) throws IOException {
```
Comment (dagnir) [inline http-clients/url-connection-client/src/main/java/software/amazon/awssdk/http/urlconnection/UrlConnectionHttpClient.java:233]: Agree we should revisit this in the future. Can you add a `TODO` here?

Comment (Bennett-Lynch) [reply http-clients/url-connection-client/src/main/java/software/amazon/awssdk/http/urlconnection/UrlConnectionHttpClient.java:233]: Will do.

**Review by dagnir — APPROVED** [review]
:shipit:

**Comment by zoewangg** [comment]
@all-contributors please add @Bennett-Lynch for code

**Pull request merged**